txnsync/bitmask.go

```diff
@@ -0,0 +1,252 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"bytes"
+	"errors"
+)
+
+var errIndexNotFound = errors.New("invalid bitmask: index not found")
+var errInvalidBitmaskType = errors.New("invalid bitmask type")
+
+//msgp:allocbound bitmask maxBitmaskSize
+type bitmask []byte
+
+// assumed to be in mode 0, sets bit at index to 1
+func (b *bitmask) setBit(index int) {
+	byteIndex := index/8 + 1
+	(*b)[byteIndex] |= 1 << (index % 8)
+}
+
+// trimBitmask compresses the bitmask into one of the 4 types:
+// type 0: input bitmask bit pos x b -> output bitmask bit pos x b
+// type 1: input bitmask bit pos x b -> output bitmask bit pos x !b
+// type 2: stores the positions of bits where the bit value is 1
+//         input bitmask first bit 1 at pos A, second bit 1 at pos B, ...
+//         output bitmask stores A, B-A, ...
+// type 3: same as type 2, but stores the positons where the bit is 0
+func (b *bitmask) trimBitmask(entries int) {
+	if *b == nil {
+		return
+	}
+	lastExists := 0
+	lastNotExists := 0
+	numExists := 0
+	for i := 0; i < entries; i++ {
+		byteIndex := i/8 + 1
+		if (*b)[byteIndex]&(1<<(i%8)) != 0 {
+			lastExists = i
+			numExists++
+		} else {
+			lastNotExists = i
+		}
+	}
+	bitmaskType := 0
+	bestSize := bytesNeededBitmask(lastExists)
+	if bestSize > bytesNeededBitmask(lastNotExists) {
+		bitmaskType = 1
+		bestSize = bytesNeededBitmask(lastNotExists)
+	}
+	if bestSize > numExists*2+1 {
+		bitmaskType = 2
+		bestSize = numExists*2 + 1
+	}
+	if bestSize > (entries-numExists)*2+1 {
+		bitmaskType = 3
+		bestSize = (entries-numExists)*2 + 1
+	}
+	switch bitmaskType {
+	case 1:
+		(*b)[0] = 1
+		for i := range *b {
+			if i != 0 {
+				(*b)[i] = 255 - (*b)[i] // invert bits
+			}
+		}
+	case 2:
+		newBitmask := make(bitmask, 1, bestSize)
+		newBitmask[0] = 2
+		last := 0
+		for i := 0; i < entries; i++ {
+			byteIndex := i/8 + 1
+			if (*b)[byteIndex]&(1<<(i%8)) != 0 {
```

Comment:
I'm not opposing for refactoring here, but I'd rather focus on correctness and performance first.

Reply:
Usually this kind of things are error-prone, so it's better to make a subroutine for correctness reasons.

Reply:
yes - the thing is that the entire section is not written with performance in mind. The common pattern here :
```golang
byteIndex := i/8 + 1
if (*b)[byteIndex]&(1<<(i%8)) != 0 {...
```
should be written as an operation where we don't recalculate the `byteIndex` and `(1<<(i%8))` on every iteration.
Moreover, I would explicitly try to loop unwrapping:
```golang
currentByte := (*b)[byteIndex]
if b & 1 != 0 {
   ...
}
if b & 2 != 0 {
   ...
}
...
```
since the amount of code here is relatively small, and could fit into a single cache line cache. Using the local variable `currentByte ` instead of the byte array ensures that the cpu doesn't need to go to it's L1 cache, but instead, allows the compiler to place this variable into one of the registers.

Reply:
It's a good point. However, I would only try to optimize it if performance of this code is really a problem. Perhaps the cost of this code is far less than network and disk, then optimizing it will only lead to more complex code. Also, there is probably a bitset library out there that is well optimized.

Reply:
This code is on the critical path - when it's slow, the transaction sync performance would get affected right away. So far ( afaik ), it was working fine - so I'm not very motivated in touching the code.. however, if we have any correctness issue, then we clearly have to fix it right away.